core/src/apps/solana/transaction/__init__.py

```diff
@@ -0,0 +1,214 @@
+        self.version: int | None = None
+
+        if serialized_tx_reader.peek() & 0b10000000:
+            self.version = serialized_tx_reader.get() & 0b01111111
+            # only version 0 is supported
+            if self.version > 0:
+                raise DataError("Unsupported transaction version")
+
+        self.required_signers_count: int = serialized_tx_reader.get()
+        self.num_signature_read_only_addresses: int = serialized_tx_reader.get()
+        self.num_read_only_addresses: int = serialized_tx_reader.get()
+
+    def _parse_addresses(self, serialized_tx_reader: BufferReader) -> None:
+        num_of_addresses = parse_var_int(serialized_tx_reader)
+
+        assert (
+            num_of_addresses
+            >= self.required_signers_count
+            + self.num_signature_read_only_addresses
+            + self.num_read_only_addresses
+        )
+
+        addresses: list[Address] = []
+        for i in range(num_of_addresses):
+            if i < self.required_signers_count:
+                type = AddressType.AddressSig
+            elif (
+                i < self.required_signers_count + self.num_signature_read_only_addresses
+            ):
+                type = AddressType.AddressSigReadOnly
+            elif (
+                i
+                < self.required_signers_count
+                + self.num_signature_read_only_addresses
+                + self.num_read_only_addresses
+            ):
+                type = AddressType.AddressRw
+            else:
+                type = AddressType.AddressReadOnly
+
+            address = parse_pubkey(serialized_tx_reader)
+
+            addresses.append((address, type))
+
+        self.addresses = addresses
+
+    def _parse_instructions(self, serialized_tx_reader: BufferReader) -> None:
+        num_of_instructions = parse_var_int(serialized_tx_reader)
+
+        instructions: list[RawInstruction] = []
+
+        for _ in range(num_of_instructions):
+            program_index = serialized_tx_reader.get()
+            program_id = base58.encode(self.addresses[program_index][0])
+            num_of_accounts = parse_var_int(serialized_tx_reader)
+            accounts: list[int] = []
+            for _ in range(num_of_accounts):
+                account_index = serialized_tx_reader.get()
+                accounts.append(account_index)
+
+            data_length = parse_var_int(serialized_tx_reader)
+
+            instruction_id_format = get_instruction_id_length(program_id)
+            instruction_id_length = instruction_id_format.length
+            # Some programs e.g. Associated Token Account Program don't include the instruction
+            # id in the data for instruction id 0 but they include it for the other instructions.
+            # Instructions with such optional instruction id also don't contain any other instruction
+            # data (otherwise parsing would be impossible).
+            if data_length < instruction_id_length:
```

Comment:
Actually no, because e.g. the Memo program only has a single instruction so it doesn't include instruction id in the data but it does contain other instruction data.

The comment was misleading so I updated it in https://github.com/trezor/trezor-firmware/pull/3359/commits/07a33b0ea99674b761bcda99186dac23047148cd and then once more in https://github.com/trezor/trezor-firmware/pull/3359/commits/5ed1636c7efc3d45f30d3752dff67fadf9a38f30.